### ISSUE TITLE
Switch from pip to pipenv or Poetry.

### ISSUE THREAD

**Issue by jenstroeger** [jenstroeger/python-package-template#1]
Currently the package uses [pip](https://github.com/pypa/pip) and Python’s [setuptools](https://github.com/pypa/setuptools) as described in [Packaging and distributing projects](https://packaging.python.org/guides/distributing-packages-using-setuptools/). However, problems with recursive dependency checks, deep version collisions, or cumbersome pinning make working with pip somewhat unreliable.

Consider switching to [pipenv](https://github.com/pypa/pipenv) or [Poetry](https://github.com/python-poetry/poetry) instead…

**Comment by jenstroeger** [comment]
Other packaging and env-management tools recommended during conversations with @hynek and @behnazh, are [flit](https://github.com/takluyver/flit) and [asdf](https://github.com/asdf-vm/asdf) and [pyenv](https://github.com/pyenv/pyenv).

**Comment by jenstroeger** [comment]
Consider also the discussion on [PEP 665: A file format to list Python dependencies for reproducibility of an application](https://discuss.python.org/t/pep-665-take-2-a-file-format-to-list-python-dependencies-for-reproducibility-of-an-application/11736).

**Comment by behnazh** [comment]
The PEP 665 makes a lot of sense. It also mentions how [npm](https://www.npmjs.com/) has solved the reproducibility problem by using [package.json](https://docs.npmjs.com/cli/v6/configuring-npm/package-json).

But for the time being Poetry seems to be doing well enough in dependency management and reproducibility I think.

**Comment by jenstroeger** [comment]
Interesting conversation [regarding `flit` on python.org](https://discuss.python.org/t/should-flit-become-a-pypa-project/11854).

**Comment by behnazh** [comment]
> Interesting conversation [regarding `flit` on python.org](https://discuss.python.org/t/should-flit-become-a-pypa-project/11854).

Interesting indeed. I didn't find any solid reasoning in the thread for Flit though. In particular, I tend to disagree with [this argument](https://discuss.python.org/t/should-flit-become-a-pypa-project/11854/5):

> my default is that any packaging-related project that wants to be part of PyPA should be welcomed. We shouldn’t need to justify including projects, rather we should require justification from people who want to say “no” to a request to join.

Welcoming any project without requiring justifications brings us to where we are now, which might be convenient for some developers but quite chaotic compared to [npm](https://www.npmjs.com/).

**Comment by jenstroeger** [comment]
So, more conversations and I’m starting to agree with @hynek’s view: all we need is the [pyptoject.toml](https://github.com/jenstroeger/python-package-template/blob/master/pyproject.toml) file to declare the dependencies within the `[build-system]` section, and if necessary pin/hash the rest.

There are more details in this [Python Application Dependency Management](https://hynek.me/articles/python-app-deps-2018/) blog post, the [setuptools docs](https://setuptools.pypa.io/en/latest/build_meta.html), and as an example we can take a look at [structlog’s pyproject.toml](https://github.com/hynek/structlog/blob/main/pyproject.toml) — which is a package, not an application, though.

**Comment by jenstroeger** [comment]
Here’s another choice by the Stripe folks, and discussion to consider: [Remove Poetry and reinstate setup.py #583](https://github.com/stripe/stripe-python/pull/583)

I’d like to know @ob-stripe’s and @koobs’ thoughts on this subject (and this issue) now, two years after that decision, and their experiences since.

**Comment by hynek** [comment]
I think you can test that yourself. Try installing pendulum (same author as Poetry, packaged w/ Poetry) into a Python 3.10 virtualenv.

**Comment by jenstroeger** [comment]
Hmm… https://github.com/regebro/pyroma 🤔

**Comment by jenstroeger** [comment]
Another interesting thread: [Should I be pinning my dependencies?](https://discuss.python.org/t/should-i-be-pinning-my-dependencies/)

TL;DR — if we’re building a package, _don’t_ pin because the package needs to live amongst other packages and their dependencies; if we’re building an app, _do_ pin and package accordingly (a somewhat neglected area of the Python ecosystem).

**Comment by behnazh** [comment]
Looking at the `Remediation` section in [Scorecard security analysis results on this repo](https://github.com/jenstroeger/python-package-template/security/code-scanning/2?query=ref%3Arefs%2Fpull%2F105%2Fmerge+ref%3Arefs%2Fpull%2F105%2Fhead+ref%3Arefs%2Fheads%2Fadd-scorecard) for pinning dependencies, its recommendation is consistent with the idea of pinning dependencies for apps and not pinning them for packages. It also references another interesting [article](https://jbeckwith.com/2019/12/18/package-lock/), which gives practical examples of the consequences of pinning dependencies in packages.

I like to look at this problem from all these perspectives:

1. supply chain security
2. reproducibility
3. conflicts in dependency resolution

As also recommended by Scorecard, checking in lock files to the repo can be used as a signature of the state of dependencies, which is essential for the first two concerns. It can, however, still not be perfect for the third issue. However, using dependency update tools, such as Dependabot can mitigate this problem when dependencies are guaranteed to be the latest, which also pass the CI tests. Or am I missing something?

I personally use `poetry` these days and this is how I use it: I keep the un-pinned dependencies in `pyproject.toml` and also check in the `poetry.lock` to have a snapshot of my dependency graph for supply chain security and reproducibility reasons.

**Comment by jenstroeger** [comment]
Interesting [comment](https://discuss.python.org/t/help-packaging-optional-application-features-using-extras/14074/7): 

> I think that in general, packaging Python _applications_ is not a well-supported scenario. The packaging ecosystem is mostly focused on packaging _libraries_. Yes, entry points allow your library to expose a command interface, but the packaging capabilities are still designed around libraries.

**Comment by jenstroeger** [comment]
Today I stumbled over the [hashin](https://github.com/peterbe/hashin) package which allows to add integrity hashes for packages, much like Node’s [package-lock.json](https://nodejs.dev/learn/the-package-lock-json-file) file.

**Comment by jenstroeger** [comment]
This conversation is growing lengthy, so I wanted to pick it up again with a few thoughts based on what we’ve learned over the past months about [packaging Python projects](https://packaging.python.org/en/latest/tutorials/packaging-projects/).

PR #74 adds Makefile support and commit https://github.com/jenstroeger/python-package-template/pull/74/commits/34b8953e24f1a000098e0377fbb276dc8f829bcd adds an interesting feature to that Makefile support: `make requirements` creates a snapshot of the Python virtual environment (aka. [`pip freeze`](https://pip.pypa.io/en/stable/cli/pip_freeze/)) including integrity hashes. Given that new Makefile support, here’s my proposal.

### Packages

If a user builds a package then specify only direct package dependencies in `setup.py` using standard [version range specifications](https://peps.python.org/pep-0508/), and install that package using e.g. `pip`.

### Applications

If a user builds an application then specify only direct package dependencies in `setup.py` using standard [version range specifications](https://peps.python.org/pep-0508/), just like for packages. In addition to that, use the new `make requirements` to pin and hash _all recursive dependencies_ for the application, and install the application using that requirements file.

**Comment by jenstroeger** [comment]
Closing this because we decided that the current approach of using `pip` and `setup.py` and `requirements.txt` is sufficient. There are issues around deep package conflict detection and resolution but that’s not anything we can/should address with this template — users may choose to use Poetry or Pipenv.
